pkg/utils/revision/revision_utils.go

```diff
@@ -0,0 +1,249 @@
+	// clone := lws.DeepCopy()
+	str := &bytes.Buffer{}
+	err := unstructured.UnstructuredJSONScheme.Encode(lws, str)
+	if err != nil {
+		return nil, err
+	}
+	patched, err := strategicpatch.StrategicMergePatch(str.Bytes(), revision.Data.Raw, lws)
+	if err != nil {
+		return nil, err
+	}
+	restoredLws := &leaderworkerset.LeaderWorkerSet{}
+	if err = json.Unmarshal(patched, restoredLws); err != nil {
+		return nil, err
+	}
+	return restoredLws, nil
+}
+
+// nextRevision finds the next valid revision number based on revisions. If the length of revisions
+// is 0 this is 1. Otherwise, it is 1 greater than the largest revision's Revision. This method
+// assumes that revisions has been sorted by Revision.
+func NextRevision(revisions []*appsv1.ControllerRevision) int64 {
+	count := len(revisions)
+	if count <= 0 {
+		return 1
+	}
+
+	max := int64(1)
+	for _, revision := range revisions {
+		if max < revision.Revision {
+			max = revision.Revision
+		}
+	}
+	return max + 1
+}
+
+// TruncateHistory cleans up all other controller revisions except the currentRevision.
+// currentRevision is the one that matches the templateHash that is passed
+func TruncateHistory(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet, templateHash string) error {
+	controllerHistory := history.NewHistory(ctx, k8sClient)
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: map[string]string{
+		leaderworkerset.SetNameLabelKey: lws.Name,
+	}})
+	if err != nil {
+		return err
+	}
+	revisions, err := controllerHistory.ListControllerRevisions(lws, selector)
+	if err != nil {
+		return err
+	}
+
+	for i, revision := range revisions {
+		if revision.Labels[leaderworkerset.TemplateRevisionHashKey] != templateHash {
+			if err := controllerHistory.DeleteControllerRevision(revisions[i]); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func EqualLeaderWorkerTemplates(lhs *leaderworkerset.LeaderWorkerSet, rhs *leaderworkerset.LeaderWorkerSet) bool {
+	if !reflect.DeepEqual(lhs.Spec.LeaderWorkerTemplate, rhs.Spec.LeaderWorkerTemplate) {
+		return false
+	}
+	if (lhs.Spec.NetworkConfig == nil || string(*lhs.Spec.NetworkConfig.SubdomainPolicy) == string(leaderworkerset.SubdomainShared)) && (rhs.Spec.NetworkConfig == nil || string(*rhs.Spec.NetworkConfig.SubdomainPolicy) == string(leaderworkerset.SubdomainShared)) {
+		return true
+	}
+
+	if lhs.Spec.NetworkConfig == nil || rhs.Spec.NetworkConfig == nil {
+		return false
+	}
+
+	return string(*lhs.Spec.NetworkConfig.SubdomainPolicy) == string(*rhs.Spec.NetworkConfig.SubdomainPolicy)
+}
+
+// Sha1Hash accepts an input string and returns the 40 character SHA1 hash digest of the input string.
+func Sha1Hash(s string) string {
+	h := sha1.New()
+	h.Write([]byte(s))
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+func LeaderWorkerTemplateHash(lws *leaderworkerset.LeaderWorkerSet) string {
```

Comment:
we don't need this anymore, we have a hash function in the history pkg now that we should converge on.

Reply:
I'm guessing you are referring to this one https://github.com/kubernetes-sigs/lws/pull/277/files#diff-356fe1c7888b3417da9a1bd377ba0eb471c7b7e97c3684a7e88165e6c14a2f14R88-R97? 

I already commented this somewhere else, but I think it also fits here: 

Don't we still need a hashing function that treats NetworkConfig == nil and NetworkConfig.SubdomainShared as the same though?

When updating from a version that doesn't have networkConfig the steps would be:
Create a controller revision where the NetworkConfig si nil. When checking if LWS was updated, it will determine that it isn't because the hash is the same.

Another change is made, similar to https://github.com/kubernetes-sigs/lws/issues/240#issuecomment-2452679190 triggering the lws webhook, and defaulting NetworkConfig. A new controller revision will be created, as the data will now be different. Now, the hash won't match the hash that is in the leaderSts, triggering a rolling update when it isn't supposed to.